Tests/MapboxCoreNavigationTests/RouteControllerTests.swift

```diff
@@ -9,538 +9,1224 @@ import MapboxNavigationNative
-            actualCoordinates.append(location.coordinate)
-            return false
-        }
-        let expectedTestCoordinatesCount = locationManager.locations.count
-        expectation(description: "All coordinates processed") {
-            actualCoordinates.count == expectedTestCoordinatesCount
+        waitForExpectations(timeout: expectationsTimeout)
+    }
+    
+    func testRerouteIfNilCustomRoutingProvider() {
+        let routeController = makeRouteController(routingProvider: nil)
+        routeController.reroute(from: rawLocation, along: routeProgress)
+        XCTAssertFalse(routeController.isRerouting)
+        XCTAssertTrue(rerouteController.forceRerouteCalled)
+    }
+    
+    func testRerouteIfFinishedRouting() {
+        routeController.finishRouting()
+        routeController.reroute(from: rawLocation, along: routeProgress)
+        XCTAssertFalse(routeController.isRerouting)
+        XCTAssertFalse(rerouteController.forceRerouteCalled)
+    }
+    
+    func testNotifyWillRerouteIfRoutingProviderFailed() {
+        let willRerouteExpectation = expectation(description: "Will reroute call on delegate")
+        delegate.onWillRerouteFrom = { location in
+            XCTAssertEqual(location, self.rawLocation)
+            willRerouteExpectation.fulfill()
         }
-
-        let replayFinished = expectation(description: "Replay Finished")
-        locationManager.replayCompletionHandler = { _ in
-            replayFinished.fulfill()
-            return false
+        
+        let didFailToReroutExpectation = expectation(description: "Did fail to reroute call on delegate")
+        delegate.onDidFailToRerouteWith = { error in
+            XCTAssertEqual(error as! DirectionsError, .unableToRoute)
+            didFailToReroutExpectation.fulfill()
         }
+        
+        routeController.reroute(from: rawLocation, along: routeProgress)
+        XCTAssertFalse(routeController.isRerouting)
+        XCTAssertFalse(rerouteController.forceRerouteCalled)
+        XCTAssertTrue(routingProvider.calculateRoutesCalled)
 
-        let speedMultiplier: TimeInterval = 100
-        locationManager.speedMultiplier = speedMultiplier
-        locationManager.startUpdatingLocation()
-
-        waitForExpectations(timeout: locationManager.expectedReplayTime, handler: nil)
-
-        let expectedCoordinates = locations.map(\.coordinate)
-        XCTAssertEqual(expectedCoordinates, actualCoordinates)
+        waitForExpectations(timeout: expectationsTimeout)
     }
-
-    func testRerouteAfterArrival() {
-        let coordinates = [
-            CLLocationCoordinate2D(latitude: 37.750384, longitude: -122.387487),
-            CLLocationCoordinate2D(latitude: 37.764343, longitude: -122.388664),
-        ]
+    
+    func testSendRerouteNotificationIfRoutingProviderFailed() {
+        let heading = CLHeading(heading: 50, accuracy: 1)!
+        routeController.locationManager(locationManagerSpy, didUpdateHeading: heading)
```

Comment:
It's because I wanted to test all the parameters in `.routeControllerWillReroute` notification and I wanted to set values for heading, because it's one of such parameters.

Now I realise that `heading` is editable so I replaced `locationManager (_:didUpdateHeading:)` call with the setter's call, it's more clear.